File: src/main/java/fs/client/ui/layout/Flex.java
package fs.client.ui.layout;

import fs.client.ui.Component;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;


public class Flex extends Component {

  private final List<Component> children = new ArrayList<>();
  private Direction direction = Direction.ROW;
  private JustifyContent justifyContent = JustifyContent.LEFT;
  private ItemAlign alignItems = ItemAlign.START;
  private ContentAlign alignContent = ContentAlign.START;
  private Wrap wrap = Wrap.NO_WRAP;

  public Flex justifyContent(JustifyContent justifyContent) {
    this.justifyContent = justifyContent;

    return this;
  }

  public Flex alignItems(ItemAlign align) {
    alignItems = align;

    return this;
  }

  public Flex direction(Direction direction) {
    this.direction = direction;

    return this;
  }

  public Flex wrap(Wrap wrap) {
    this.wrap = wrap;

    return this;
  }

  public Flex alignContent(ContentAlign align) {
    alignContent = align;

    return this;
  }

  @Override
  public int preferredWidth() {
    int preferredWidth = 0;
    for (Component component : children) {
      preferredWidth += component.preferredWidth();
    }
    return preferredWidth;
  }

  @Override
  public int preferredHeight() {
    int preferredHeight = 0;
    for (Component component : children) {
      preferredHeight += component.preferredHeight();
    }
    return preferredHeight;
  }

  public Flex bounds(int x, int y, int width, int height) {
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;

    return this;
  }

  @Override
  public void render() {
    int mainAxisSize = direction.main(width, height);
    List<List<Component>> componentRows = wrap.split(children, direction, width, height);

    int usedCrossSpace = 0;
    for (List<Component> componentRow : componentRows) {
      int rowCrossSpace = 0;
      for (Component component : componentRow) {
        int componentCrossSpace = direction.cross(component.preferredWidth(), component.preferredHeight());
        if (rowCrossSpace < componentCrossSpace) {
          rowCrossSpace = componentCrossSpace;
        }
      }
      usedCrossSpace += rowCrossSpace;
    }
    int crossAxisSpace = direction.cross(width, height);
    int extraCrossSpace = crossAxisSpace - usedCrossSpace;

    int componentCrossOffset = 0;
    int betweenCrossOffset = 0;
    int additionalCrossSpace = 0;
    if (alignContent == ContentAlign.START) {
      componentCrossOffset = 0;
      betweenCrossOffset = 0;
      additionalCrossSpace = 0;
    } else if (alignContent == ContentAlign.END) {
      componentCrossOffset = extraCrossSpace;
      betweenCrossOffset = 0;
      additionalCrossSpace = 0;
    } else if (alignContent == ContentAlign.CENTER) {
      componentCrossOffset = extraCrossSpace / 2;
      betweenCrossOffset = 0;
      additionalCrossSpace = 0;
    } else if (alignContent == ContentAlign.SPACE_BETWEEN) {
      componentCrossOffset = 0;
      betweenCrossOffset = componentRows.size() <= 1 ? 0 : extraCrossSpace / (componentRows.size() - 1);
      additionalCrossSpace = 0;
    } else if (alignContent == ContentAlign.SPACE_AROUND) {
      betweenCrossOffset = extraCrossSpace / componentRows.size();
      componentCrossOffset = betweenCrossOffset / 2;
      additionalCrossSpace = 0;
    } else if (alignContent == ContentAlign.STRETCH) {
      componentCrossOffset = 0;
      betweenCrossOffset = 0;
      additionalCrossSpace = extraCrossSpace / componentRows.size();
    }

    for (List<Component> componentRow : componentRows) {
      int rowMainSpace = 0;
      int rowCrossSpace = 0;
      for (Component component : componentRow) {
        rowMainSpace += direction.main(component.preferredWidth(), component.preferredHeight());
        int componentCrossSpace = direction.cross(component.preferredWidth(), component.preferredHeight());
        if (rowCrossSpace < componentCrossSpace) {
          rowCrossSpace = componentCrossSpace;
        }
      }
      rowCrossSpace += additionalCrossSpace;

      int extraMainSpace = mainAxisSize < rowMainSpace ? 0 : mainAxisSize - rowMainSpace;
      float mainAxisCoefficient = (rowMainSpace < mainAxisSize) ? 1 : (float) mainAxisSize / (float) rowMainSpace;

      int initialMainOffset = 0;
      int betweenMainOffset = 0;
      if (justifyContent == JustifyContent.LEFT) {
        initialMainOffset = 0;
        betweenMainOffset = 0;
      } else if (justifyContent == JustifyContent.RIGHT) {
        initialMainOffset = extraMainSpace;
        betweenMainOffset = 0;
      } else if (justifyContent == JustifyContent.CENTER) {
        initialMainOffset = extraMainSpace / 2;
        betweenMainOffset = 0;
      } else if (justifyContent == JustifyContent.SPACE_BETWEEN) {
        initialMainOffset = 0;
        betweenMainOffset = componentRow.size() <= 1 ? 0 : extraMainSpace / (componentRow.size() - 1);
      } else if (justifyContent == JustifyContent.SPACE_AROUND) {
        betweenMainOffset = extraMainSpace / componentRow.size();
        initialMainOffset = betweenMainOffset / 2;
      }

      int componentMainOffset = initialMainOffset;
      int maxCrossSpace = 0;
      Iterator<Component> componentIterator = direction.iterator(componentRow);
      while (componentIterator.hasNext()) {
        Component component = componentIterator.next();
        int componentMainSpace = (int) (direction.main(component.preferredWidth(), component.preferredHeight()) * mainAxisCoefficient);
        int componentCrossSpace = direction.cross(component.preferredWidth(), component.preferredHeight());

        int initialCrossOffset = 0;
        int cross = 0;
        if (alignItems == ItemAlign.START) {
          cross = componentCrossSpace;
          initialCrossOffset = 0;
        } else if (alignItems == ItemAlign.END) {
          cross = componentCrossSpace;
          initialCrossOffset = rowCrossSpace - cross;
        } else if (alignItems == ItemAlign.CENTER) {
          cross = componentCrossSpace;
          initialCrossOffset = (rowCrossSpace - cross) / 2;
        } else if (alignItems == ItemAlign.STRETCH) {
          cross = rowCrossSpace;
          initialCrossOffset = 0;
        }

        component.bounds(
            x + direction.x(componentMainOffset, componentCrossOffset + initialCrossOffset),
            y + direction.y(componentMainOffset, componentCrossOffset + initialCrossOffset),
            direction.x(componentMainSpace, cross),
            direction.y(componentMainSpace, cross)
        ).render();

        componentMainOffset += componentMainSpace + betweenMainOffset;
        if (maxCrossSpace < cross) {
          maxCrossSpace = cross;
        }
      }

      componentCrossOffset += maxCrossSpace + betweenCrossOffset;
    }
  }

  @Override
  public Component findComponentAt(int x, int y) {
    if (this.x <= x && x <= this.x + this.width) {
      if (this.y <= y && y <= this.y + this.height) {
        for (Component child : children()) {
          Component childComponent = child.findComponentAt(x, y);
          if (childComponent != null) {
            return childComponent;
          }
        }
      }
    }

    return null;
  }

  @Override
  public List<Component> children() {
    return children;
  }

  public Flex add(Component child) {
    child.parent(this);
    children.add(child);

    return this;
  }

  public enum Direction {
    ROW {
      public int main(int x, int y) {
        return x;
      }

      public int cross(int x, int y) {
        return y;
      }

      @Override
      public int x(int main, int cross) {
        return main;
      }

      @Override
      public int y(int main, int cross) {
        return cross;
      }

      @Override
      public Iterator<Component> iterator(List<Component> componentRow) {
        return componentRow.iterator();
      }
    },
    ROW_REVERSE {
      public int main(int x, int y) {
        return x;
      }

      public int cross(int x, int y) {
        return y;
      }

      @Override
      public int x(int main, int cross) {
        return main;
      }

      @Override
      public int y(int main, int cross) {
        return cross;
      }

      @Override
      public Iterator<Component> iterator(List<Component> componentRow) {
        List<Component> reverse = new ArrayList<>(componentRow);
        Collections.reverse(reverse);
        return reverse.iterator();
      }
    },
    COLUMN {
      public int main(int x, int y) {
        return y;
      }

      public int cross(int x, int y) {
        return x;
      }

      @Override
      public int x(int main, int cross) {
        return cross;
      }

      @Override
      public int y(int main, int cross) {
        return main;
      }

      @Override
      public Iterator<Component> iterator(List<Component> componentRow) {
        return componentRow.iterator();
      }
    },
    COLUMN_RESVERSE {
      public int main(int x, int y) {
        return y;
      }

      public int cross(int x, int y) {
        return x;
      }

      @Override
      public int x(int main, int cross) {
        return cross;
      }

      @Override
      public int y(int main, int cross) {
        return main;
      }

      @Override
      public Iterator<Component> iterator(List<Component> componentRow) {
        List<Component> reverse = new ArrayList<>(componentRow);
        Collections.reverse(reverse);
        return reverse.iterator();
      }
    };

    public abstract int main(int x, int y);

    public abstract int cross(int x, int y);

    public abstract int x(int main, int cross);

    public abstract int y(int main, int cross);

    public abstract Iterator<Component> iterator(List<Component> componentRow);
  }

  public enum Wrap {
    WRAP {
      public List<List<Component>> split(List<Component> components, Direction direction, int width, int height) {
        int mainAxisSize = direction.main(width, height);
        List<List<Component>> componentRows = new ArrayList<>();
        int currentComponentIndex = 0;
        while (currentComponentIndex < components.size()) {
          List<Component> componentRow = new ArrayList<>();

          int cumulativeMainAxisSize = 0;
          while (currentComponentIndex < components.size()) {
            Component component = components.get(currentComponentIndex);
            int componentMainSize = direction.main(component.preferredWidth(), component.preferredHeight());
            if (cumulativeMainAxisSize + componentMainSize > mainAxisSize) {
              break;
            } else {
              cumulativeMainAxisSize += componentMainSize;
              componentRow.add(component);
              currentComponentIndex++;
            }
          }
          componentRows.add(componentRow);
        }

        return componentRows;
      }
    },
    NO_WRAP {
      public List<List<Component>> split(List<Component> components, Direction direction, int width, int height) {
        List<List<Component>> componentRows = new ArrayList<>();
        componentRows.add(components);
        return componentRows;
      }
    };

    public abstract List<List<Component>> split(List<Component> components, Direction direction, int width, int height);
  }

  public enum JustifyContent {
    LEFT,
    CENTER,
    RIGHT,
    SPACE_BETWEEN,
    SPACE_AROUND
  }

  public enum ItemAlign {
    START,
    END,
    CENTER,
    STRETCH
  }


  public enum ContentAlign {
    START,
    END,
    CENTER,
    STRETCH,
    SPACE_BETWEEN,
    SPACE_AROUND
  }
}

File: src/main/kotlin/com/fracturedskies/render/components/Flex.kt
package com.fracturedskies.render.components

import com.fracturedskies.engine.collections.Context
import com.fracturedskies.engine.collections.Key
import com.fracturedskies.engine.jeact.AbstractComponent
import com.fracturedskies.engine.jeact.Bounds
import com.fracturedskies.engine.jeact.Component
import com.fracturedskies.engine.jeact.Node
import java.util.*


class Flex(attributes: Context) : AbstractComponent<Unit>(attributes, Unit) {
  companion object {
    val DIRECTION = Key<Direction>("direction")
    val JUSTIFY_CONTENT = Key<JustifyContent>("justifyContent")
    val ALIGN_ITEMS = Key<ItemAlign>("itemAlign")
    val ALIGN_CONTENT = Key<ContentAlign>("contentAlign")
    val WRAP = Key<Wrap>("wrap")
    fun Node.Builder<*>.flex(
            direction: Direction = Direction.ROW,
            justifyContent: JustifyContent = JustifyContent.LEFT,
            alignItems: ItemAlign = ItemAlign.START,
            alignContent: ContentAlign = ContentAlign.START,
            wrap: Wrap = Wrap.WRAP,
            block: Node.Builder<Unit>.() -> (Unit) = {}
    ) {
      nodes.add(Node(::Flex, Context(
              DIRECTION to direction,
              JUSTIFY_CONTENT to justifyContent,
              ALIGN_ITEMS to alignItems,
              ALIGN_CONTENT to alignContent,
              WRAP to wrap
      ), block))
    }
  }

  private val direction get() = requireNotNull(attributes[DIRECTION])
  private val justifyContent get() = requireNotNull(attributes[JUSTIFY_CONTENT])
  private val alignItems get() = requireNotNull(attributes[ALIGN_ITEMS])
  private val alignContent get() = requireNotNull(attributes[ALIGN_CONTENT])
  private val wrap get() = requireNotNull(attributes[WRAP])

  override fun preferredWidth(): Int {
    return children.sumBy({ it.preferredWidth() })
  }

  override fun preferredHeight(): Int {
    return children.sumBy({ it.preferredHeight() })
  }

  override fun render(bounds: Bounds) {
    this.bounds = bounds

    val mainAxisSize = direction.main(bounds.width, bounds.height)
    val componentRows = wrap.split(children, direction, bounds.width, bounds.height)

    var usedCrossSpace = 0
    for (componentRow in componentRows) {
      val rowCrossSpace = componentRow
              .map({ direction.cross(it.preferredWidth(), it.preferredHeight()) })
              .max()
              ?: 0
      usedCrossSpace += rowCrossSpace
    }
    val crossAxisSpace = direction.cross(bounds.width, bounds.height)
    val extraCrossSpace = crossAxisSpace - usedCrossSpace

    var componentCrossOffset = alignContent.componentCrossOffset(componentRows, extraCrossSpace)
    val betweenCrossOffset = alignContent.betweenCrossOffset(componentRows, extraCrossSpace)
    val additionalCrossSpace = alignContent.additionalCrossSpace(componentRows, extraCrossSpace)
    for (componentRow in componentRows) {
      var rowMainSpace = 0
      var rowCrossSpace = 0
      for (component in componentRow) {
        rowMainSpace += direction.main(component.preferredWidth(), component.preferredHeight())
        val componentCrossSpace = direction.cross(component.preferredWidth(), component.preferredHeight())
        if (rowCrossSpace < componentCrossSpace) {
          rowCrossSpace = componentCrossSpace
        }
      }
      rowCrossSpace += additionalCrossSpace

      val extraMainSpace = if (mainAxisSize < rowMainSpace) 0 else mainAxisSize - rowMainSpace
      val mainAxisCoefficient = if (rowMainSpace < mainAxisSize) 1f else mainAxisSize.toFloat() / rowMainSpace.toFloat()
      val initialMainOffset = justifyContent.initialOffset(componentRow, extraMainSpace)
      val betweenMainOffset = justifyContent.betweenOffset(componentRow, extraMainSpace)

      var componentMainOffset = initialMainOffset
      var maxCrossSpace = 0
      val componentIterator = direction.iterator(componentRow)
      while (componentIterator.hasNext()) {
        val component = componentIterator.next()
        val componentMainSpace = (direction.main(component.preferredWidth(), component.preferredHeight()) * mainAxisCoefficient).toInt()
        val componentCrossSpace = direction.cross(component.preferredWidth(), component.preferredHeight())

        val initialCrossOffset = alignItems.offset(componentCrossSpace, rowCrossSpace)
        val cross = alignItems.cross(componentCrossSpace, rowCrossSpace)

        component.render(Bounds(
                bounds.x + direction.x(componentMainOffset, componentCrossOffset + initialCrossOffset),
                bounds.y + direction.y(componentMainOffset, componentCrossOffset + initialCrossOffset),
                direction.x(componentMainSpace, cross),
                direction.y(componentMainSpace, cross)
        ))

        componentMainOffset += componentMainSpace + betweenMainOffset
        if (maxCrossSpace < cross) {
          maxCrossSpace = cross
        }
      }

      componentCrossOffset += maxCrossSpace + betweenCrossOffset
    }
  }
}

enum class Direction {
  ROW {
    override fun main(x: Int, y: Int) = x
    override fun cross(x: Int, y: Int) = y
    override fun x(main: Int, cross: Int) = main
    override fun y(main: Int, cross: Int) = cross
    override fun iterator(componentRow: List<Component<*>>) = componentRow.iterator()
  },
  ROW_REVERSE {
    override fun main(x: Int, y: Int) = x
    override fun cross(x: Int, y: Int) = y
    override fun x(main: Int, cross: Int) = main
    override fun y(main: Int, cross: Int) = cross
    override fun iterator(componentRow: List<Component<*>>) = componentRow.reversed().iterator()
  },
  COLUMN {
    override fun main(x: Int, y: Int) = y
    override fun cross(x: Int, y: Int) = x
    override fun x(main: Int, cross: Int) = cross
    override fun y(main: Int, cross: Int) = main
    override fun iterator(componentRow: List<Component<*>>) = componentRow.iterator()
  },
  COLUMN_RESVERSE {
    override fun main(x: Int, y: Int) = y
    override fun cross(x: Int, y: Int) = x
    override fun x(main: Int, cross: Int) = cross
    override fun y(main: Int, cross: Int) = main
    override fun iterator(componentRow: List<Component<*>>) = componentRow.reversed().iterator()
  };

  abstract fun main(x: Int, y: Int): Int
  abstract fun cross(x: Int, y: Int): Int
  abstract fun x(main: Int, cross: Int): Int
  abstract fun y(main: Int, cross: Int): Int
  abstract fun iterator(componentRow: List<Component<*>>): Iterator<Component<*>>
}

enum class Wrap {
  WRAP {
    override fun split(components: List<Component<*>>, direction: Direction, width: Int, height: Int): List<List<Component<*>>> {
      val mainAxisSize = direction.main(width, height)
      val componentRows = ArrayList<List<Component<*>>>()
      var currentComponentIndex = 0
      while (currentComponentIndex < components.size) {
        val componentRow = mutableListOf<Component<*>>()

        var cumulativeMainAxisSize = 0
        while (currentComponentIndex < components.size) {
          val component = components[currentComponentIndex]
          val componentMainSize = direction.main(component.preferredWidth(), component.preferredHeight())
          if (cumulativeMainAxisSize + componentMainSize > mainAxisSize && componentRow.size > 0) {
            break
          } else {
            cumulativeMainAxisSize += componentMainSize
            componentRow.add(component)
            currentComponentIndex++
          }
        }
        componentRows.add(componentRow)
      }

      return componentRows
    }
  },
  NO_WRAP {
    override fun split(components: List<Component<*>>, direction: Direction, width: Int, height: Int): List<List<Component<*>>> {
      val componentRows = ArrayList<List<Component<*>>>()
      componentRows.add(components)
      return componentRows
    }
  };

  abstract fun split(components: List<Component<*>>, direction: Direction, width: Int, height: Int): List<List<Component<*>>>
}

enum class JustifyContent {
  LEFT {
    override fun initialOffset(componentRow: List<Component<*>>, extraMainSpace: Int) = 0
    override fun betweenOffset(componentRow: List<Component<*>>, extraMainSpace: Int) = 0
  },
  CENTER {
    override fun initialOffset(componentRow: List<Component<*>>, extraMainSpace: Int) = extraMainSpace / 2
    override fun betweenOffset(componentRow: List<Component<*>>, extraMainSpace: Int) = 0
  },
  RIGHT {
    override fun initialOffset(componentRow: List<Component<*>>, extraMainSpace: Int) = extraMainSpace
    override fun betweenOffset(componentRow: List<Component<*>>, extraMainSpace: Int) = 0
  },
  SPACE_BETWEEN {
    override fun initialOffset(componentRow: List<Component<*>>, extraMainSpace: Int) = 0
    override fun betweenOffset(componentRow: List<Component<*>>, extraMainSpace: Int): Int {
      return if (componentRow.size <= 1) 0 else extraMainSpace / (componentRow.size - 1)
    }
  },
  SPACE_AROUND {
    override fun initialOffset(componentRow: List<Component<*>>, extraMainSpace: Int) = betweenOffset(componentRow, extraMainSpace) / 2
    override fun betweenOffset(componentRow: List<Component<*>>, extraMainSpace: Int) = extraMainSpace / componentRow.size
  };

  abstract fun initialOffset(componentRow: List<Component<*>>, extraMainSpace: Int): Int
  abstract fun betweenOffset(componentRow: List<Component<*>>, extraMainSpace: Int): Int
}

enum class ItemAlign {
  START {
    override fun cross(columnCrossSpace: Int, rowCrossSpace: Int) = columnCrossSpace
    override fun offset(columnCrossSpace: Int, rowCrossSpace: Int) = 0
  },
  END{
    override fun cross(columnCrossSpace: Int, rowCrossSpace: Int) = columnCrossSpace
    override fun offset(columnCrossSpace: Int, rowCrossSpace: Int) = rowCrossSpace - columnCrossSpace
  },
  CENTER{
    override fun cross(columnCrossSpace: Int, rowCrossSpace: Int) = columnCrossSpace
    override fun offset(columnCrossSpace: Int, rowCrossSpace: Int) = (rowCrossSpace - columnCrossSpace) / 2
  },
  STRETCH{
    override fun cross(columnCrossSpace: Int, rowCrossSpace: Int) = rowCrossSpace
    override fun offset(columnCrossSpace: Int, rowCrossSpace: Int) = 0
  };

  abstract fun cross(columnCrossSpace: Int, rowCrossSpace: Int): Int
  abstract fun offset(columnCrossSpace: Int, rowCrossSpace: Int): Int
}

enum class ContentAlign {
  START {
    override fun componentCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
    override fun betweenCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
    override fun additionalCrossSpace(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
  },
  END {
    override fun componentCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = extraCrossSpace
    override fun betweenCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
    override fun additionalCrossSpace(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
  },
  CENTER {
    override fun componentCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = extraCrossSpace / 2
    override fun betweenCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
    override fun additionalCrossSpace(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
  },
  STRETCH {
    override fun componentCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
    override fun betweenCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
    override fun additionalCrossSpace(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) =
            extraCrossSpace / componentRows.size
  },
  SPACE_BETWEEN {
    override fun componentCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
    override fun betweenCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) =
            if (componentRows.size <= 1) 0 else extraCrossSpace / (componentRows.size - 1)
    override fun additionalCrossSpace(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
  },
  SPACE_AROUND {
    override fun componentCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) =
            betweenCrossOffset(componentRows, extraCrossSpace) / 2
    override fun betweenCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) =
            extraCrossSpace / componentRows.size
    override fun additionalCrossSpace(componentRows: List<List<Component<*>>>, extraCrossSpace: Int) = 0
  };

  abstract fun componentCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int): Int
  abstract fun betweenCrossOffset(componentRows: List<List<Component<*>>>, extraCrossSpace: Int): Int
  abstract fun additionalCrossSpace(componentRows: List<List<Component<*>>>, extraCrossSpace: Int): Int
}